WIP TS Drop Pod Reinforcements

KanoYugoro opened the pull request:
Make test, and make check come back clear as of right now.

It is a little bit hacky, as the drop pods themselves don't actually fire the weapon, but it sort of does work.  I think with some adjustments to a couple variables for animation offsets, it will be indistinguishable from the original in TS.

I'm opening this PR before it's ready to be merged for two reasons.  The first being that I'm not sure I went about implementing this in a way that's in line with the other support powers, and would like some direction from someone more knowledgeable about the engine.  It's entirely possible I did a workaround to get some logic I didn't know existed elsewhere or couldn't find.

Second, I'm having a logic error around the creation of the units / drop pods that I can't seem to debug.  The first unit in the list is ignored and it is not spawned / dropped.

Any and all help / review is appreciated.  Thanks in advance.

KanoYugoro committed:
Drop pod superweapon added

KanoYugoro committed:
Allow Nullable ActorLostNotification

To "mute" the unit lost notification if it is null.  This sort of
replicates part of the Insignificant ini flags from the original games.

reaperrr commented on OpenRA.Mods.Cnc/Traits/SupportPowers/DropPodsPower.cs:
```diff
@@ -0,0 +1,158 @@
+	{
+		[Desc("DropPod Unit")]
+		[ActorReference(typeof(AircraftInfo))]
+		public readonly string UnitType = "badr";
```
Speaking from experience, setting the internal default to null and forcing modders to set the unit type explicitly via `[FieldLoader.Require]` causes the least problems, especially in case this ever needs an update rule.

reaperrr commented on OpenRA.Mods.Cnc/Traits/SupportPowers/DropPodsPower.cs:
```diff
@@ -0,0 +1,158 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2018 The OpenRA Developers (see AUTHORS)
```
-2019.

reaperrr commented on OpenRA.Mods.Cnc/OpenRA.Mods.Cnc.csproj:
```diff
@@ -73,6 +73,7 @@
     <Compile Include="Traits\Render\WithDeliveryAnimation.cs" />
     <Compile Include="Traits\Render\WithRoof.cs" />
     <Compile Include="Traits\SupportPowers\IonCannonPower.cs" />
+	<Compile Include="Traits\SupportPowers\DropPodsPower.cs" />
```
Unlike .yaml files, csproj should use spaces instead of tabs.

reaperrr commented on OpenRA.Mods.Common/Traits/Sound/ActorLostNotification.cs:
```diff
@@ -33,8 +33,11 @@ public ActorLostNotification(ActorLostNotificationInfo info)
 		{
-			var player = info.NotifyAll ? self.World.LocalPlayer : self.Owner;
-			Game.Sound.PlayNotification(self.World.Map.Rules, player, "Speech", info.Notification, self.Owner.Faction.InternalName);
+			if (info.Notification != null && info.Notification != "")
```
`if (!string.IsNullOrEmpty(info.Notification))`

reaperrr commented on mods/ts/audio/speech-generic.yaml:
```diff
@@ -26,6 +26,7 @@ Speech:
 		CriticalStructureLost: 00-i196
 		CriticalUnitLost: 00-i194
 		DisablePower: 00-i230
+        DropPodsAvailable: 00-i506
```
While the engine supports spaces in yaml, all our official mods use tabs, so please make sure to fix all places that use spaces.
If you have trouble identifying them locally, you can see it easily here on gh (red = tabs, no red = spaces).

reaperrr commented on mods/ts/rules/gdi-structures.yaml:
```diff
@@ -433,14 +433,13 @@ GAPLUG:
 		Footprint: xxx xxx
 		Dimensions: 2,3
 	WithIdleOverlay@DISH:
-		RequiresCondition: !build-incomplete
```
This looks like a rebase error, we don't want the overlays to show and animate before the buildup is complete.
Please revert this.

reaperrr commented on mods/ts/rules/gdi-structures.yaml:
```diff
@@ -433,14 +433,13 @@ GAPLUG:
 		Sequence: idle-dish
 	WithIdleOverlay@LIGHTS:
-		RequiresCondition: !build-incomplete && !disabled && !empdisable
+		RequiresCondition: !disabled && !empdisable
```
Same here...

reaperrr commented on mods/ts/rules/gdi-structures.yaml:
```diff
@@ -433,14 +433,13 @@ GAPLUG:
 		Sequence: idle-lights
 	WithIdleOverlay@STRIP:
-		RequiresCondition: !build-incomplete && !disabled && !empdisable
+		RequiresCondition: !disabled && !empdisable
```
...and here.

KanoYugoro committed:
Address reaperrr's feedback

Replaced leading spaces with tabs wherever I could find them, renamed
UnitType to DropPodType, Re-Added accidentally deleted RequiresConditions,
added in a cleaner null check in ActorLostNotification, and updated
CopyRight year.

KanoYugoro commented:
Thank you for the review, I appreciate the feedback.  I tried to address everything you raised in that last commit.

teinarss commented on mods/ts/rules/gdi-structures.yaml:
```diff
@@ -488,37 +502,52 @@ GAPLUG:
+		Sequence: idle-droppodb
 	ProvidesPrerequisite@buildingname:
-	SelectionDecorations:
+	SelectionDecorations:
```
Missing a newline at the end. See https://unix.stackexchange.com/questions/18743/whats-the-point-in-adding-a-new-line-to-the-end-of-a-file?fbclid=IwAR1iuN8Luca-Y0FJh_CdNRt2eo5zTAHnSg4-SxFlNqywrT1shu3NHV0lgtc

teinarss commented on mods/ts/rules/gdi-support.yaml:
```diff
@@ -236,3 +236,19 @@ GAPLUG3:
+	Plug:
+		Type: plug.droppod
+	Power:
+		Amount: -20
```
Same here!

teinarss commented on mods/ts/sequences/aircraft.yaml:
```diff
@@ -28,3 +28,9 @@ apache:
+	Inherits: ^VehicleOverlays
+	idle:
+		Facings: 1
+		Length: 1
```
And here.

teinarss commented on mods/ts/weapons/explosions.yaml:
```diff
@@ -66,3 +66,8 @@ Demolish:
+DropPodExplode:
+	Warhead@1Eff: CreateEffect
+		Explosions: droppod_explosion
+		ExplosionPalette: effect-ignore-lighting-alpha75
```
And the last one.

pchote commented on OpenRA.Mods.Cnc/Traits/SupportPowers/DropPodsPower.cs:
```diff
@@ -0,0 +1,159 @@
+				var targetCell = self.World.Map.CellContaining(target);
+				var PodLocations = self.World.Map.FindTilesInCircle(targetCell, info.PodScatter).Shuffle(self.World.SharedRandom);
+
+				using (var pe = PodLocations.GetEnumerator())
```
Please use a `foreach` here for consistency.

pchote commented on OpenRA.Mods.Cnc/Traits/SupportPowers/DropPodsPower.cs:
```diff
@@ -0,0 +1,159 @@
+					foreach (var u in units)
+					{
+						CPos podDropCellPos = pe.Current;
+						w.Add(new IonCannon(self.Owner, info.WeaponInfo, w, self.World.Map.CenterOfCell(podDropCellPos), Target.FromCell(w, podDropCellPos),
```
I suspect that we shouldn't be using `IonCannon` for this.

pchote commented:
Any progress on @teinarss's requests here @KanoYugoro?

pchote reviewed (commented):
I haven't tested this yet, but from looking at the code I have a few comments of my own:

reaperrr commented:
There haven't been any updates, closing and moving to Future milestone.
